### PR TITLE
[Friesian] Add Unittest for RecallInitializer & FeatureInitializer

### PR DESCRIPTION
**Pull request by hoshibara**
### 1. Why the change?

A part of [Friesian Serving unittest](https://github.com/orgs/analytics-zoo/projects/10)

### 3. Summary of the change

* add Unittest for FeatureInitializer
* add Unittest for RecallInitializer

### 4. How to test?

* [x] Unit test
* [x] GitHub workflow

### 5. New dependencies

* friesian pom add dependencies:
  * org.junit.jupiter.junit-jupyter - license: EPL 2.0, scope test, Copyright © 2022 The JUnit Team
* friesian pom add plugins:
  * maven-surefire-plugin - license: Apache 2.0, Copyright © 2004-2015 The Apache Software Foundation.
  * maven-failsafe-plugin - license: Apache 2.0, Copyright © 2004-2015 The Apache Software Foundation.
* parent pom add properties:
  * junit5-maven-surfire-plugin.version
  * junit5-maven-failsafe-plugin.version
  * junit-jupyter.version

### CODE REVIEW

File: scala/friesian/src/test/java/FeatureInitializerTest.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class FeatureInitializerTest {
+
+
+    public static Seq<String> convertListToSeq(String[] inputArray) {
+        return JavaConverters.asScalaIteratorConverter(Arrays.asList(inputArray).iterator()).asScala().toSeq();
+    }
+
+    String generateID(String keyPrefix, String ID) {
```
Comment (cyita) [inline scala/friesian/src/test/java/FeatureInitializerTest.java:27]: Please use MGet(String keyPrefix, String[] keys) in LettuceUtils.

File: scala/friesian/src/test/java/FeatureInitializerTest.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        } else {
+            return redisKeyPrefix + keyPrefix + ":" + ID;
+        }
+    }
+
+    void checkRedisRecord(LettuceUtils redis, String keyPrefix, Dataset<Row> dataset) {
+        dataset.collectAsList().forEach((row) -> {
+            Object[] rowData = (Object[]) EncodeUtils.bytesToObj(
+                    Base64.getDecoder().decode(
+                            redis.getSync().getdel(generateID(keyPrefix, row.get(0).toString()))));
```
Comment (cyita) [inline scala/friesian/src/test/java/FeatureInitializerTest.java:46]: Please use `MGet`.

File: scala/friesian/src/test/java/RecallInitializerTest.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import java.util.Map;
+
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class RecallInitializerTest {
+
+    @Test
+    void testInitialization() throws IOException, InterruptedException {
+        String configPath = "/home/xingyuan/projects/serving/BigDL/scala/friesian/src/test/resources/nearlineConfig/config_recall.yaml";
```
Comment (cyita) [inline scala/friesian/src/test/java/RecallInitializerTest.java:22]: Please use resource path.

File: scala/friesian/src/test/java/RecallInitializerTest.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        // you can get initialDataPath file from friesian-serving.tar.gz
+
+        SparkSession sparkSession = SparkSession.builder().getOrCreate();
+        Dataset<Row> dataset = sparkSession.read().parquet(NearlineUtils.helper().getInitialDataPath());
+
+        IndexService indexService = new IndexService(NearlineUtils.helper().getIndexDim());
+        indexService.load(NearlineUtils.helper().getIndexPath());
+
+        assertEquals(dataset.count(), indexService.getNTotal());
+        Map<Integer, double[]> vectorMap = new HashMap<>();
```
Comment (cyita) [inline scala/friesian/src/test/java/RecallInitializerTest.java:33]: Check NTotal is enough.

File: scala/friesian/src/test/java/FeatureInitializerTest.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+                assertEquals(rowData[i].toString(), row.get(1 + i).toString());
+            }
+//            assertNotEquals(redis.getSync().getdel(generateID(keyPrefix, row.get(0).toString())), null);
+        });
+    }
+
+    @Test
+    void testInitialization() throws IOException, InterruptedException {
+        String configPath = "/home/xingyuan/projects/serving/BigDL/scala/friesian/src/test/resources/nearlineConfig/config_feature_vec.yaml";
+        FeatureInitializer.main(new String[]{"-c", configPath});
```
Comment (cyita) [inline scala/friesian/src/test/java/FeatureInitializerTest.java:57]: Flushall before loading data.

File: scala/friesian/src/test/java/RecallInitializerTest.java (hunk 4 above)
Comment (hoshibara) [reply scala/friesian/src/test/java/RecallInitializerTest.java]: In the following code, there is assertArrayEquals for inserted vectors.

File: .github/workflows/nb-ut-scala-friesian.yml [hunk 6]
```diff
@@ -11,7 +11,7 @@ on:
 jobs:
   build:
 
-    runs-on: [ self-hosted, Gondolin, ubuntu-20.04-lts, CLX, AVX512, Ettenmoors ]
+    runs-on: [ self-hosted, Gondolin, ubuntu-20.04, CLX, AVX512, Ettenmoors ]
```
Comment (cyita) [inline .github/workflows/nb-ut-scala-friesian.yml:14]: `CLX,ubuntu-20.04,AVX512,Ettenmoors`

File: scala/friesian/src/main/java/com/intel/analytics/bigdl/friesian/serving/feature/utils/LettuceUtils.java [hunk 7]
```diff
@@ -125,10 +127,9 @@ private LettuceUtils(RedisType redisType, ArrayList<Tuple2<String, Integer>> red
     public static LettuceUtils getInstance(RedisType redisType, ArrayList<Tuple2<String, Integer>> redisHostPort,
                                            String redisPrefix, String sentinelMasterURL, String sentinelMasterName,
                                            int itemSlotType) {
-        if (instance == null) {
-            instance = new LettuceUtils(redisType, redisHostPort, redisPrefix, sentinelMasterURL,
-                    sentinelMasterName, itemSlotType);
-        }
+        // TODO: Thread-Safe
```
Comment (cyita) [inline scala/friesian/src/main/java/com/intel/analytics/bigdl/friesian/serving/feature/utils/LettuceUtils.java:130]: Please revert changes here.

File: scala/pom.xml [hunk 8]
```diff
@@ -136,7 +136,7 @@
 
 
         <!-- Plugin versions and config -->
-        <maven-surefire-plugin.version>2.20.1</maven-surefire-plugin.version>
+        <maven-surefire-plugin.version>2.22.2</maven-surefire-plugin.version>
```
Comment (cyita) [inline scala/pom.xml:139]: Please check the version with dongjie.

File: scala/friesian/src/test/resources/testConfig/config_friesian.yaml [hunk 9]
```diff
@@ -0,0 +1 @@
+dataFramePath: /tmp/friesian-unittest/share/data1.parquet
```
Comment (cyita) [inline scala/friesian/src/test/resources/testConfig/config_friesian.yaml:1]: Please delete this file.

**Pull request merged**